### DISCUSSION TITLE
Broadcasting to multiple client with websocket

### DISCUSSION THREAD

**Discussion by pmbanugo** [ergo-services/ergo#200] [General]
How do I broadcast to a group (or all) connected websocket client? The example shows replaying message to the same client. I tried to figure out how to broadcast message to all connected client but I couldn't figure it out, and the examples doesn't do that.

**Comment by halturin** [comment]
that example starts 3 processes to handle all the incoming web socket connections. you can use `Events` feature, but you still need to implement broadcasting to the websocket-connections.

**Reply by pmbanugo** [reply]
I couldn't find a means to get hold of the websocket connections. I tried to use the Alias/PID but it seems to only direct the message to the last  client that connected to that pool. Here's what my current code looks like: [GitHub Gist](https://gist.github.com/pmbanugo/7537e824ee2ce07a4507e5f204e0f872). are my missing something?

**Comment by halturin** [comment]
I would design it this way:

- spawn a process that acts as a publisher (register and send event)
- spawn a set of ws worker processes (monitor/link the event)
- in a ws worker process keep the aliases of ws connections (just a list or a map)

the flow looks like: publisher sends an event (`gen.Process.SendEvent`), all ws workers receive this event (`gen.MessageEvent`) and transform the data, say, into the json, send this json to all the ws clients in their list/map (`gen.Process.Send` or `gen.Process.SendAlias`)

**Reply by halturin** [reply]
probably better to make this transformation on the publisher side (if all the ws clients are expecting the same json format). in this case, ws process has only to forward the data as it is (but as a `websocket.Message`).

**Reply by pmbanugo** [reply]
Thanks for the design suggestion. Maybe because this is my first Go program that I'm struggling to figure this out.

I can't find a way to get access to the actual ws connection. I tried to use the ID (Alias) from when the message was sent for subscription, but it doesn't work as expected, it just sends it to the last ws connection/client. So I'm guessing that Alias (message.ID) is a reference to the meta-process, not the actual WS connection. I can't find a way to get access to the ws connection.

Can you point me to the docs or is there a method besides HandleMessage that can give me access to that? Or is the ws connection process the value of gen.PID that sent the message to the ws worker?

**Reply by halturin** [reply]
the main idea of actor model is message passing. meta-processes serves those ws connections so you dont need to have access to. Just send websocket.Message to ID (Alias) of meta-process that handles ws connection. and handle incoming messages in HandleMessage callback of your ws worker process. Thats it. 

If you need to close ws connection, just send `exit` signal to the meta-process using `gen.Process.SendExit(...)`

https://docs.ergo.services/extra-library/meta-processes/websocket

**Reply by halturin** [reply]
here you can read more about the meta-processes https://docs.ergo.services/basics/meta-process
